### PR TITLE
GUA-747 - Agree and document architecture for the Activity Log

### PR DESCRIPTION
**Pull request by huwd**
Outlines and documents the proposed approach, paths not taken and other considerations.

Largely a repeat of what we did with the service card store in the end.

### CODE REVIEW

File: docs/adr/0006-recording-activity-log-data-within-the-account.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+### Service record data structure
+
+```typescript
+type UrnFdnSub = string;
+type ClientId = string;
+
+export interface UserData {
+  user_id: UrnFdnSub;
+  govuk_signin_journey_id: string;
```
Comment (divyaparameswaran) [inline docs/adr/0006-recording-activity-log-data-within-the-account.md:114]: Should we include the session_id here - https://github.com/alphagov/di-account-management-backend/pull/95/files#diff-3d97f46fb56381153aaf510597651dc011f9911b6bb87e3f380207992e986ec8R8

File: docs/adr/images/2023-004-ActivityLog.png
[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/images/2023-004-ActivityLog.png:1]: I believe it was https://cloudcraft.co
Sounds like we're gonna need to find a better tool pretty soon, but that's worked for now.

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/images/2023-004-ActivityLog.png:1]: As for doing the diagram,
Yeah, it's mostly that the labels of what it'll export here aren't accurate.

I think the ideal for me is to move this stuff to something like mermaid, something markdown that can meet accessibility better.

And then have a separate file after that perhaps generates us an overall diagram either by conversing with our IoC or perhaps with AWS itself. Need to look into what makes sense.

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/images/2023-004-ActivityLog.png:1]: But for now, I think so long as we four are clear on the components, let's merge this and upgrade later?

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/images/2023-004-ActivityLog.png:1]: You happy to resolve this @DougsCER ?

**Review by btarquini — COMMENTED** [review]
Questions from security:
Are the 2 boxes in 1 AWS account?
Is the design below wrapped in a custom VPC(s)?
Are the Lambda's tied to the VPC (so any ingress/egress is through the VPC)
Are you using SQS endpoints?
Are DynamoDB's communicating via VPC endpoints?
How is data being captured from TxMA? Is this data over the internet or over the AWS backbone?
Is the Frontend connected via Fargate (a similar design can be found here (https://govukverify.atlassian.net/wiki/spaces/DFA/pages/3520233478/Data+flow+-+Infrastructure) and if so, are we using Private/Public Subnets?
Are CMK's stored in AWS Secrets Manager?
What is the rotation period for these keys?
Are we using WAF for the front end with Shield Advanced?
Is envelope encryption being used? (as from my understanding, that's what other teams are doing when handling TxMA data) - https://docs.aws.amazon.com/wellarchitected/latest/financial-services-industry-lens/use-envelope-encryption-with-customer-master-keys.html
If the design is based upon 2 other existing solutions, has an ITHC been conducted against them?

File: docs/adr/0006-recording-activity-log-data-within-the-account.md
[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/0006-recording-activity-log-data-within-the-account.md:1]: >  Are the Lambda's tied to the VPC (so any ingress/egress is through the VPC)
> Are you using SQS endpoints?
> Are DynamoDB's communicating via VPC endpoints?

I'll need to get back to you on these ones not really sure what we mean here, but I suspect Karol can put me on the right track

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/0006-recording-activity-log-data-within-the-account.md:1]: > Is the Frontend connected via Fargate (a similar design can be found here (https://govukverify.atlassian.net/wiki/spaces/DFA/pages/3520233478/Data+flow+-+Infrastructure) 
 
Broadly yes, however there are some differences at the start of the stack, see [this comment here](https://github.com/alphagov/di-account-management-frontend/blob/main/deploy/template.yaml#L8)

I think I'm not best placed to evaluate exactly where these are similar or differ however everything for these is in [this file](https://github.com/alphagov/di-account-management-frontend/blob/main/deploy/template.yaml#L8). We may need to convene the right combination of SRE / Security / folks who can read cloudformation between the two teams for a proper comparison?
 
 > and if so, are we using Private/Public Subnets?
 
 [Yes](https://github.com/alphagov/di-account-management-frontend/blob/main/deploy/template.yaml#L767-L768), though we're using the Protected subnets as defined [here](https://govukverify.atlassian.net/wiki/spaces/PLAT/pages/3531735041/VPC)

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/0006-recording-activity-log-data-within-the-account.md:1]: >  Are CMK's stored in AWS Secrets Manager?

I'm keen not to misspeak here so here's all our [KMS key config](https://github.com/alphagov/di-account-management-backend/blob/d7bca45b9f993cf0b31f2461fe89ef69a0ea59b7/infrastructure/template.yaml#L1574-L1769), can you see the answers you need from that? 

I've had a look and can't see these keys in secrets manager, could we unpack a bit more on the requirement there?

I suspect some of our answers may be in what we're doing with [these docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-kms-key.html#cfn-kms-key-enablekeyrotation).
.
>  What is the rotation period for these keys?

As in the cloudformation they should all have `EnableKeyRotation: true`. Looking at the docs [for that](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-kms-key.html#cfn-kms-key-enablekeyrotation), seems we're looking at:
> one year after the enable date and every year thereafter

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/0006-recording-activity-log-data-within-the-account.md:1]: >  Are we using WAF for the front end with Shield Advanced?

We are using [WAF](https://github.com/alphagov/di-account-management-frontend/blob/5bee31653b97c6fcb1bc7483a0d55e4a42a8afb0/deploy/template.yaml#L8) yes. I've gone in to check AWS shield in integration and it's not set up.

However I believe we do have alternative DDoS protection via CloudFront

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/0006-recording-activity-log-data-within-the-account.md:1]: > Is envelope encryption being used? (as from my understanding, that's what other teams are doing when handling TxMA data) - https://docs.aws.amazon.com/wellarchitected/latest/financial-services-industry-lens/use-envelope-encryption-with-customer-master-keys.html

Not yet, that's next up.

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/0006-recording-activity-log-data-within-the-account.md:1]: >  If the design is based upon 2 other existing solutions, has an ITHC been conducted against them?

Nope, these solutions were made before I think we had a process setup.

[the remark this answers is not in the document]
Comment (huwd) [reply docs/adr/0006-recording-activity-log-data-within-the-account.md:1]: Right these ones, I've been trying to work out how to give you a good answer all afternoon.
I reckon this is likely tied to [ADR 33](https://github.com/alphagov/digital-identity-architecture/blob/main/adr/0033-egress-controls-for-core-apps.md), right?

I think the answer there is "we are trying to use standard dev platform components".
This becomes the domain of [@andyloughran](https://github.com/orgs/alphagov/people/andyloughran) and I think he's best placed to get you an answer. I'll follow up there with something in slack.

**Review by divyaparameswaran — APPROVED** [review]

**Pull request merged**